### ISSUE TITLE
On a mac running engine-q get Command k to work.

### ISSUE THREAD

**Issue by stormasm** [nushell/reedline#196]
In nushell when I type Command-k it clears the screen and we are happy...

In engine-q it clears the screen and then 
on the next keystroke
it takes me back to where I was in the terminal
when I typed Command-k

It is remembering where I last was and goes back there...

It would be great to have this working as I use it all the time...

**Comment by sholderbach** [comment]
Can you try running the reedline demo binary with `cargo run -- -k` to see if we receive the `Command-k`

According to this comment the `Command-k` implementation seems to be driven by the terminal emulator. 
https://superuser.com/a/819605/1646405

reedline implements `Ctrl-L` for screen clearing (only form feed, not clearing the scrollback buffer)

Solution would be to intercept the `Command-k` and resetting the cursor position accordingly

**Comment by stormasm** [comment]
@sholderbach ---- interesting ---- when I run the reedline demo as you suggested I am seeing NO ISSUES with Command-k.  It works perfectly with no issue...

So something is getting in the way between this command

```rust
cargo run -- -k
```

and engine-q....  

I guess this is really good news, now we need to chase down the culprit....

**Comment by stormasm** [comment]
So here is another important point.  The next day I went back to this terminal and closed out of the reedline demo...  And now for some reason that terminal is acting strange once I exit.  I have to kill that terminal completely...  So something is still amiss (unfortunately)...

**Comment by sholderbach** [comment]
Do you have a screenshot of how it is acting up? Might be a case of raw mode not being correctly disabled, that reedline registered some phantom resizes during sleep etc. Darren had some issues after resuming https://discord.com/channels/601130461678272522/855886335980994600/913788532269658152 where I don't think we found the culprit yet

`cargo run -- -k` should just print the key events and shouldn't respond to other events. How was the behavior with just running `reedline`

**Comment by stormasm** [comment]
@sholderbach 
Ok lets start over again on my comments from yesterday...
I was WRONG in my diagnosis...

Everything works fine with just running reedline...

In order to get out of it I have to kill the process....
And that is where I got confused...

I killed the process yesterday..

This morning I sat down and my terminal was messed up.
Then I sent you the message...

So obviously killing the process will wreak havoc on everything...
And things will not reset properly...

So in conclusion we are working fine with reedline no issues...

The only issue now is when we run reedling inside engine-q which is not working...
reedline standalone works fine.
sorry for the confusion.

**Comment by sholderbach** [comment]
> So obviously killing the process will wreak havoc on everything...
> And things will not reset properly...

This sounds like the terminal remains in raw mode afterwards. Also happens if reedline should panic for other reasons

We only guard that for recoverable Errors

https://github.com/nushell/reedline/blob/a44a590b25e2fb2593db24a17c06ed887633a5fc/src/engine.rs#L299-L312

**Comment by sholderbach** [comment]
@stormasm Can you give this a try again? I made a little change in #164 to the prompt initialisation step. Maybe this resolves this?

**Comment by fdncred** [comment]
for what it's worth, on my Mac, with e-q, I hit command-k and it clears the terminal screen. however, the `>` is at the top of the terminal but the cursor is at the bottom. so when I start typing it starts on the bottom line of the terminal but the `>` is at the top.
<img width="586" alt="Screen Shot 2021-12-23 at 2 18 07 PM" src="https://user-images.githubusercontent.com/343840/147288191-dcff3432-e5df-4c7b-adf6-c84eec94bd02.png">

**Comment by sholderbach** [comment]
Right, we don't register that... The arrow probably is result of it printing starting wherever and the buffer is painted at the last remembered position. Best case solution would be to register the command k as a key press. worst case solution would be to check for a change in cursor position on every repaint.

**Comment by fdncred** [comment]
when command-k happens move cursor to 0,0 may fix it

**Comment by stormasm** [comment]
@sholderbach ok so here is where we are at with this.... I just tested it running the reedline demo.  Thanks for your work on this and your persistence in looking into this issue !!

on my mac 
CTRL L appears to be working now...
I am pretty sure it is fixed...
CMD K is still broken

I remember @jntrnr saying something to me in two weeks ago conference call
that it might be as simple as just adding the key binding in engine-q for
the CMD-k sequence on the mac or just in general.

maybe all we have to do is map the CMD-K to what CTRL-L is doing ?

@fdncred might know what the difference is between CTRL-L and CMD-K
on the mac is --- at the moment my brain is not telling me the difference...

**Comment by fdncred** [comment]
The problem with all this keybinding mess is that whatever terminal each of us is using can override the keybindings. Also, whatever shell we're using can override the keybindings. So, CTRL-L for you may not be CTRL-L for me. 

So, in the Tabby terminal app on Mac, using fish, CTRL-L does nothing. CMD-K clears the screen.

In mac's default terminal app, using bash or fish, CTRL-L clears the screen and so does CMD-K.

In mac's default terminal app, using Nushell, CTRL-L does nothing but CMD-K clears the screen. I don't have any keybindings in Nushell so it must be using whatever the defaults are.

I've used iterm2 and kitty in the past and they may behave differently. I'm not sure. I'm currently using Tabby as my default terminal app on Mac. What are you using @stormasm?

**Comment by stormasm** [comment]
@fdncred so I am using https://en.wikipedia.org/wiki/Terminal_(macOS)

https://support.apple.com/guide/terminal/welcome/mac

Its just the standard terminal that ships with the mac....

When I was running my demo to test this stuff I was running the
straight reedline demo code, I was not in engine-q...

**Comment by stormasm** [comment]
@fdncred so this is good news (IMHO) because on my mac CTRL L now clears the screen correctly with no issues.  Before there was some weirdness (I believe)...

@sholderbach based on what @fdncred said in his recent comment, it appears then "all we have to do" is in the default reedline implementation simply make CTRL L and CMD K do the same thing ?  This way folks on a default mac using its terminal will experience the same behavior as they are currently seeing in nushell where CMD K just works like it is "supposed to"...

Thoughts on this ?  Again thank you for checking this out...

**Comment by fdncred** [comment]
BTW @stormasm I found this that explains the differences between the two key combinations. https://superuser.com/questions/819593/what-is-the-difference-between-ctrll-and-cmdk-for-clearing-the-terminal-screen

**Comment by sholderbach** [comment]
Can you try to confirm that crossterm is able to register the keypress cmd-k? If the terminal emulator were to completely intercept it we wouldn't be able to respond to it directly but would instead have to react to the fact that the cursor likely moved to the start of the screen between the last repaint and now? 
I was trying to find some developer documentation from Apple on that or if grepping the zsh source would let something jump out at me but I don't know if there is a separate signal.

**Comment by stormasm** [comment]
@sholderbach 
I am not sure the best way to find this out, I don't think they
have a discord or news group associated with crossterm.

You may know as you have been working with crossterm for awhile
via reedline...

So we could ask the developers the question directly....
Via email ??

Or we could file an issue directly ??

I think you might be the better person to send the email or ask
the question
as I am not sure I fully understand the question you are asking
or the issue exactly --- and I think you probably understand it 
better than I...

Timon doesn't have an email...
https://github.com/TimonPost

https://github.com/Canop
cano.petrole@gmail.com

Canop recently made a commit to the repo so he might know..
But Timon is the main developer.

My email is stormasm@gmail.com
If you want to Cc: me on it...

What would be the best way for us to find this out ?
No hurry --- and not sure the best way to go...

But I did a google search and don't see the answer either.
So I just wanted to respond back.

Thank you...

**Comment by fdncred** [comment]
Apple's terminal app appears to intercept CMD-K but CTRL-L is heard by reedline. I can tell this by running `cargo run -- -k`

**Comment by stormasm** [comment]
@fdncred thank you, that makes sense...

**Comment by stormasm** [comment]
@sholderbach @fdncred ok so I ran the 

```rust
cargo run -- -k
```
on my mac with the mac terminal (the one referenced above)

it registers the CTRL-L and displays like @fdncred above...

But when I type the CMD-k it clears the screen
@fdncred I guess you are seeing this too as stated above...
And there is no message showing up...

So that is why in reedline or engine-q when you type CMD-k
reedline takes you back to the spot you were at before you
typed the CMD-k...  Because it never knows it was even typed
as @sholderbach mentions above...  Now I understand 👍 

Both of you might also be interested to note that I ran

https://github.com/alacritty/alacritty

And guess what ?

They display the 

Char: l code: 0x00006c; Modifier CONTROL; Flags 0b000010

when I type CMD-k

so we are seeing the same EXACT output for both 
CTRL-L and CMD-K with alacritty

so there you have it :)

Finally we understand what is going on...  This is great news
we finally got it...  Thanks to both of you for figuring this out...

**Comment by fdncred** [comment]
@stormasm you may also be able to change the keybindings in the terminal app `ctrl-,`. I've messed with those in the past.

**Comment by stormasm** [comment]
@fdncred thank you !  and I happy you agree with closing this issue for now is OK.
